### PR TITLE
Feedback will now act like a TransferFunction

### PR DESCRIPTION
**Pull request by abhiphile**
#### Brief description of what is fixed or changed
Feedback will now act like a TransferFunction.
Solves Issue https://github.com/sympy/sympy/issues/26161

#### Release Notes

* physics.control
  * Added support for using Feedback as a TransferFunction.

### CODE REVIEW

File: sympy/physics/control/lti.py
Comment (VascoSch92) [inline sympy/physics/control/lti.py:1141]: why `if(isinstance(other, Feedback)):` and not just if `isinstance(other, Feedback):`  as you did before?

Comment (abhiphile) [reply sympy/physics/control/lti.py]: Thankyou for pointing it out. I've corrected it .

File: sympy/physics/control/lti.py
Comment (sylee957) [inline sympy/physics/control/lti.py:2328]: I may still rethink about inheriting `Feedback` from `TransferFunction`.

Comment (abhiphile) [reply sympy/physics/control/lti.py]: Yes it should be from `TransferFunction`. I've done the changes.

File: sympy/physics/control/lti.py
Comment (moorepants) [inline sympy/physics/control/lti.py:2308]: The current design does not ever call doit() internally. The user is left to do that as needed. If the numerator and denominator could be Series or Parallel objects, then we doit() would not be needed internally.

Comment (abhishek-kuma) [reply sympy/physics/control/lti.py:2308]: Yes I think it is not needed here , all tests are passing without it.

File: sympy/physics/control/lti.py
Comment (abhiphile) [inline sympy/physics/control/lti.py:1142]: Here `doit()` is needed because otherwise it gives an error while multiplying a `Feedback` object to a `TransferFunction` object.

Comment (abhiphile) [reply sympy/physics/control/lti.py:1142]: Am I not supposed to not use `doit()` here also ?

Comment (faze-geek) [reply sympy/physics/control/lti.py:1142]: I don't think` doit() `should be used internally at any place.

File: sympy/physics/control/lti.py
Comment (faze-geek) [inline sympy/physics/control/lti.py:2214]: I'm not sure about it. Also making changes to single input single output systems but not modifying multiple input multiple output systems seem odd and confusing to me.
https://github.com/sympy/sympy/blob/46a41c0d706f75cf44a920e2a2fd0f2e8b8f8298/sympy/physics/control/lti.py#L2503-L2506

Comment (abhiphile) [reply sympy/physics/control/lti.py:2214]: For MIMO systems feedback we have to use ` TransferFunctionMatrix` as parent class , but according to https://github.com/sympy/sympy/issues/26161 this says only to implement `Feedback` as a `TransferFunction`.

Comment (abhiphile) [reply sympy/physics/control/lti.py:2214]: I've tried using `TransferFunctionMatrix` as a parent class of `MIMOFeedback` but there it shows it as undefined 
image
```
>>> from sympy.physics.control.lti import *
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "E:\Sympy\sympy\sympy\physics\control\__init__.py", line 1, in <module>
    from .lti import (TransferFunction, Series, MIMOSeries, Parallel, MIMOParallel,
  File "E:\Sympy\sympy\sympy\physics\control\lti.py", line 2537, in <module>
    class MIMOFeedback(TransferFunctionMatrix):
                       ^^^^^^^^^^^^^^^^^^^^^^
NameError: name 'TransferFunctionMatrix' is not defined. Did you mean: 'TransferFunction'?
```
Also it is not the problem of the IDE

File: sympy/physics/control/lti.py
Comment (abhiphile) [reply sympy/physics/control/lti.py]: `doit()` for `Feedback` should not be used but for other objects it is already used eg:
https://github.com/sympy/sympy/blob/101e6d0215b5ada022577f1b77e512e54f24b59e/sympy/physics/control/lti.py#L2465

**Comment by abhiphile** [comment]
@faze-geek can you please check this PR, and suggest future work here ?

File: sympy/physics/control/lti.py
Comment (faze-geek) [reply sympy/physics/control/lti.py:2248]: Unresolving this. I would like this be addressed before we think about merging.

**Comment by faze-geek** [comment]
> Changing both `Feedback` and `MIMOFeedback` in a single PR will be messy;

I'm not fully convinced about the design decision we would take with `MIMOFeedback` at this moment. Till then you can change `Feedback` in this PR itself.

File: sympy/physics/control/lti.py
Comment (faze-geek) [inline sympy/physics/control/lti.py:1096]: Try to make lines like these simpler and more readable. It will be helpful to future users.

File: sympy/physics/control/tests/test_lti.py
Comment (faze-geek) [inline sympy/physics/control/tests/test_lti.py:1684]: The changes you made to operations of Feedback like `add`, `sub`, `mul` and `div` don't reflect here. Can you cover those in your unit tests ?

**Comment by faze-geek** [comment]
I have not reviewed any changes to `MIMOFeedback` here, go ahead and make a new PR for it.

File: sympy/physics/control/tests/test_lti.py
Comment (abhiphile) [reply sympy/physics/control/tests/test_lti.py]: Yes , Thanks for mentioning I've added tests that cover these test cases also.

File: sympy/physics/control/lti.py
Comment (abhiphile) [reply sympy/physics/control/lti.py]: I've changed these lines to a simple if else statement in the updated commit.

**Comment by faze-geek** [comment]
It's good practice to give unique commit names relevant to the changes in that commit. Keep that in mind for future.

**Comment by abhiphile** [comment]
> It's good practice to give unique commit names relevant to the changes in that commit. Keep that in mind for future.

Thank you for the guidance! I'll make sure to provide more descriptive commit names in the future.

**Comment by abhiphile** [comment]
@faze-geek @moorepants Sir, can you please review this PR? So that I could also work on `MIMOFeedback`. I think all the issues are resolved here.

**Review by faze-geek — APPROVED** [review]
The logic looks okay to me, I'd wait for others to make sure that the design is perfect before merging.

File: sympy/physics/control/tests/test_lti.py
Comment (faze-geek) [inline sympy/physics/control/tests/test_lti.py:1699]: Hi, just for proof of correctness at my end can you point me to any links where these tests have been picked up from. I usually added MATLAB examples as my unit tests.

**Comment by faze-geek** [comment]
I will merge this by tomorrow EOD if reviewers don't have any other concerns.

File: sympy/physics/control/tests/test_lti.py
Comment (abhiphile) [reply sympy/physics/control/tests/test_lti.py:1699]: Ohk, Sir I'll include the tests from MATLAB or other genuine sources and Update the tests by today.

File: sympy/physics/control/tests/test_lti.py
Comment (faze-geek) [inline sympy/physics/control/tests/test_lti.py:1694]: Should we not actually return the numerator, returning a `TransferFunction` here will be useless as what the user wants with `.num` or `.den` is- 

```
>>> fd1 = Feedback(TransferFunction(s + 1, 1 , s), TransferFunction(s + 2, 1, s))
>>> fd1
Feedback(TransferFunction(s + 1, 1, s), TransferFunction(s + 2, 1, s), -1)
>>> fd1 = fd1.doit()
>>> fd1
TransferFunction(s + 1, (s + 1)*(s + 2) + 1, s)
>>> fd1.num
s + 1
>>> fd1.den
(s + 1)*(s + 2) + 1
```

File: sympy/physics/control/tests/test_lti.py
Comment (faze-geek) [reply sympy/physics/control/tests/test_lti.py:1699]: It's okay if you don't find relevant MATLAB examples, I would still try to use other online or textbook examples rather than making up tests on my own.

Comment (faze-geek) [reply sympy/physics/control/tests/test_lti.py:1699]: I have gone over the tests, they look good to me!

File: sympy/physics/control/tests/test_lti.py
Comment (faze-geek) [reply sympy/physics/control/tests/test_lti.py:1694]: I think we can merge, once this is addressed.

Comment (abhiphile) [reply sympy/physics/control/tests/test_lti.py:1694]: Yes, sir, I've included it in tests because in the original issue message here https://github.com/sympy/sympy/issues/26161#issue-2114036253
 It says that we can also call the .num and. den in the Feedback.

Comment (abhiphile) [reply sympy/physics/control/tests/test_lti.py:1694]: > I'd also expect things like Feedback.num and Feedback.den to work. So can Feedback be a subclass of a TransferFunction

Comment (faze-geek) [reply sympy/physics/control/tests/test_lti.py:1694]: Yes, I agree. The numerator should be the `symbolic expression` (numerator of the equivalent transfer function) instead of the `TransferFunction` is what I meant.

Comment (faze-geek) [reply sympy/physics/control/tests/test_lti.py:1694]: I would like to have this behavior.
```
assert num == s + 1
assert den == (s + 1)*(s + 2) + 1
```

Comment (abhiphile) [reply sympy/physics/control/tests/test_lti.py]: I've added two properties to `Feedback` num and den.
I think, from this, it will follow this behaviour.

File: sympy/physics/control/lti.py
Comment (faze-geek) [inline sympy/physics/control/lti.py:2448]: I would say the first line is self-explanatory. Remove this.
Same for den.

Comment (abhiphile) [reply sympy/physics/control/lti.py]: I've removed the extra lines.

**Comment by moorepants** [comment]
I see this was merged, but I still don't understand why we are calling `.doit()` in library code for controls. As I understand, we originally designed these things to retain the independent form of all blocks in any given arrangement of transfer functions. For example, you should be able to create a closed loop system made of several transfer functions and then "print" the block diagram. If you call doit() internally, you'll never be able to do that.

Feedback should look like:

```python
f = Feedback(tf1, tf2)
f.num == tf1
f.dem == Parallell(1, Series(tf1, tf2))
```

no?

**Comment by abhiphile** [comment]
> I see this was merged, but I still don't understand why we are calling `.doit()` 
Sir, In this review comment it was said that the `num` and `den` should be expr which is why I've used it.

**Comment by moorepants** [comment]
@faze-geek we can either automatically evaluate these transfer function or not. If we do the former, you might as well just work with an expression and there is no need for a TransferFunction object, right? I don't quite understand the design decisions.

A sympy expression is a tree, for example $(a^2 + bx)/y$ is:

```
Mul(Add(Pow(a, 2), Mul(b, x)), Pow(y, -1))
```

The user never sees this and the printer displays `a**2 + b*x`.

We originally designed `TransferFunction` to follow a similar behavior using `TransferFunction`, `Parallel`, `Series` as corollaries to `Pow`, `Add`, and `Mul`. This retains the structure and elements that make up a block diagram of transfer functions. If we called `.doit()` internally, we lose that structure and might as well use normal SymPy `Expr` objects. So I don't understand why we are merging these types of changes.

**Comment by faze-geek** [comment]
I actually was thinking that using `.doit()` is wrong in code and was shy to merge but thought this is indeed what the issue demanded. Let's revert this for now and discuss more on the issue.

**Comment by moorepants** [comment]
It isn't that is wrong, but we do all need to understand what the design is and it needs to be written down in the docs so that people follow the design pattern.

**Comment by abhiphile** [comment]
Sir, I've understood the design and requirements, and I'm currently working on this.

**Pull request merged**
